From 6a3d68b56fb89a7c571413baea35a68c69e15980 Mon Sep 17 00:00:00 2001
From: Brian Delwiche <delwiche@google.com>
Date: Fri, 11 Oct 2024 21:06:29 +0000
Subject: [PATCH] Fix type confusion in avdt_msg.cc

It is possible for a malicious user to reply to a pending AVDT message
with the wrong type, leading to type confusion and an eventual OOB
access.

Add message type validation.

Bug: 273995284
Test: m libbluetooth
Ignore-AOSP-First: security
Tag: #security
(cherry picked from https://googleplex-android-review.googlesource.com/q/commit:12d26748f31d855ac18de9df1832f60a4bd73265)
Merged-In: I87a0df92710658d6c27d9b63ee7813a0d45a301a
Change-Id: I87a0df92710658d6c27d9b63ee7813a0d45a301a
---
 system/stack/avdt/avdt_msg.cc | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/system/stack/avdt/avdt_msg.cc b/system/stack/avdt/avdt_msg.cc
index e6286b8ada..d52dc8e500 100644
--- a/system/stack/avdt/avdt_msg.cc
+++ b/system/stack/avdt/avdt_msg.cc
@@ -1586,6 +1586,14 @@ void avdt_msg_ind(AvdtpCcb* p_ccb, BT_HDR* p_buf) {
         avdt_msg_send_grej(p_ccb, sig, &msg);
       }
     }
+
+    /* validate reject/response against cached sig */
+    if (((msg_type == AVDT_MSG_TYPE_RSP) || (msg_type == AVDT_MSG_TYPE_REJ)) &&
+        (p_ccb->p_curr_cmd == nullptr || p_ccb->p_curr_cmd->event != sig)) {
+      AVDT_TRACE_WARNING(
+          "Dropping msg with mismatched sig; sig=%d", sig);
+      ok = false;
+    }
   }
 
   if (ok && !gen_rej) {
-- 
2.48.1.262.g85cc9f2d1e-goog

